src/Foundatio.Aliyun/Storage/AliyunFileStorage.cs

```diff
@@ -0,0 +1,199 @@
+        }
+
+        public void Dispose() { }
+
+        private bool IsNotFoundException(Exception ex) {
+            if (ex is AggregateException aggregateException) {
+                foreach (var innerException in aggregateException.InnerExceptions) {
+                    if (IsNotFoundException(innerException)) {
+                        return true;
+                    }
+                }
+            }
+            if (ex is WebException webException && webException.Response is HttpWebResponse response) {
+                return response.StatusCode == HttpStatusCode.NotFound;
+            }
+            return false;
+        }
+
+        private string NormalizePath(string path) {
+            return path?.Replace('\\', '/');
+        }
+
+        public async Task<Stream> GetFileStreamAsync(string path, CancellationToken cancellationToken = default(CancellationToken)) {
+            if (string.IsNullOrEmpty(path)) {
+                throw new ArgumentNullException(nameof(path));
+            }
+            var response = await Task.Factory.FromAsync(
+                (request, callback, state) => _client.BeginGetObject(request, callback, state),
+                result => _client.EndGetObject(result),
+                new GetObjectRequest(_bucketName, NormalizePath(path)), null);
+            return response.Content;
+        }
+
+        public Task<FileSpec> GetFileInfoAsync(string path) {
+            if (string.IsNullOrEmpty(path)) {
+                throw new ArgumentNullException(nameof(path));
+            }
+            path = NormalizePath(path);
+            try {
+                var metadata = _client.GetObjectMetadata(_bucketName, path);
+                return Task.FromResult(new FileSpec {
+                    Path = path,
+                    Size = metadata.ContentLength,
+                    Created = metadata.LastModified,
+                    Modified = metadata.LastModified
+                });
+            }
+            catch (Exception) {
+                return Task.FromResult((FileSpec)null);
+            }
+        }
+
+        public Task<bool> ExistsAsync(string path) {
+            if (string.IsNullOrEmpty(path)) {
+                throw new ArgumentNullException(nameof(path));
+            }
+            try {
+                return Task.FromResult(_client.DoesObjectExist(_bucketName, NormalizePath(path)));
+            }
+            catch (Exception ex) when (IsNotFoundException(ex)) {
+                return Task.FromResult(false);
+            }
+        }
+
+        public async Task<bool> SaveFileAsync(string path, Stream stream, CancellationToken cancellationToken = default(CancellationToken)) {
+            if (string.IsNullOrEmpty(path)) {
+                throw new ArgumentNullException(nameof(path));
+            }
+            if (!stream.CanSeek) {
+                var memory = new MemoryStream();
+                await stream.CopyToAsync(memory);

```

Comment:
Any awaited async operation needs to call .AnyContext() otherwise a deadlock could occur :) Is there any reason we can't just pass the stream straight to the upload? I think it would be assumed that if you pass a stream to a save method that the stream would be read completely and modified.

Reply:
In the OssClient, when saving files with PutObject method, it will call the stream.Length property. But when the stream cannot be seek, there will be exception thrown. Please refer to the unit test https://github.com/FoundatioFx/Foundatio/blob/c092b26d42f1bd3fc7a09aaca48054b92c92a5fa/test/Foundatio.TestHarness/Storage/FileStorageTestsBase.cs#L165
So we have to copy the stream to memory to support stream that cannot be seek. Although there will be memory allocation. Do you have any better suggestion?